docs/src/crypto-primitives/tsmt.md

```diff
@@ -0,0 +1,123 @@
+                /                       \
+               /                         \
+              /                           \
+             /                             \
+            /                               \
+           /                                 \
+          /                                   \
+         /                                     \
+        /                                       \
+       /                                         \
+      /                                           \
+     /                                             \
+    /                                               \
+   /                                                 \
+  -                                                   -
+ / \                                                 / \
+o   -                                               -   o
+   / \                                             / \
+  o   -                                           -   o
+     / \                                         / \
+    o   -                                       -   o
+       / \                                     / \
+      o   -                                   -   o
+         / \                                 / \
+        o   -                               -   o
+           / \                             / \
+          o   -                           -   o
+             / \                         / \
+            o   -                       -   o
+               / \                     / \
+              o   -                   -   o
+                 / \                 / \
+                o   -               -   o
+                   / \             / \
+                  o   -           -   o
+                     / \         / \
+                    o   -       -   o
+                       / \     / \
+                      o   -   o   -
+                         / \     / \
+                        A   B   C   D
+```
+where `o` signifies an empty sub-tree. This gets compacted to:
+
+```
+          R'
+         / \
+        /   \
+       /     \
+      -       -
+     / \     / \
+    A   B   C   D 
+```
+
+Compaction implies then that the tree depth is logarithmic in the size of the dictionary entries as opposed to the size of the key space. This in turn means that the number of hashing required for executing operations on the key-value map scales logarithmically in the size of the map.
+
+The first documented implementation of a compact SMT was described [here](https://github.com/proofchains/python-proofmarshal/blob/master/proofmarshal/merbinnertree.py). This implementation relies on augmenting nodes with
+a field called `prefix` which is the longest common prefix of all childs of a given node. A disadvantage of this implementation, as well as all known implementations of compact SMTs, is the reliance of their operations on bit-wise manipulations. This makes them very expensive to implement inside Miden VM.
```

Comment:
Could we add a bit more detail on how prefixing works (e.g., that hash of the leaf now contains both the key and the value). I think it is OK to expand this part somewhat - even if it becomes two paragraphs.

Reply:
Yes, I have elaborated more. Let me know if more elaboration is needed.